control/parse.go

```diff
@@ -34,30 +34,62 @@ import (
 )
 
 // A Paragraph is a block of RFC2822-like key value pairs. This struct contains
-// two methods to fetch values, a Map called Values, and a Slice called
+// two methods to fetch values, a Map called values, and a Slice called
 // Order, which maintains the ordering as defined in the RFC2822-like block
+// The map called values contains lowercased keys. You should access it via
+// methods Get() and Set() of Pararagraph.
 type Paragraph struct {
-	Values map[string]string
+	values map[string]string
 	Order  []string
 }
 
 // Paragraph Helpers {{{
+func NewParagraph() Paragraph {
+	return Paragraph{
+		Order:  []string{},
+		values: map[string]string{},
+	}
+}
+
+// This is dangerous function, but sometimes it is a nice shortcut.
+func (p Paragraph) GetValues() map[string]string {
+	return p.values
+}
 
 func (p *Paragraph) Set(key, value string) {
-	if _, found := p.Values[key]; found {
+	k := strings.ToLower(key)
+	if _, found := p.values[k]; found {
 		/* We've got the key */
-		p.Values[key] = value
+		p.values[k] = value
 		return
 	}
 	/* Otherwise, go ahead and set it in the order and dict,
 	 * and call it a day */
 	p.Order = append(p.Order, key)
-	p.Values[key] = value
+	p.values[k] = value
+}
+
+func (p Paragraph) Get2(key string) (string, bool) {
```

Comment:
The name `Get2` is funny. Maybe let's only provide this form of `Get`, and other uses can do `value, _ = p.Get` or make their own wrapper. If we want to add the other one later, that seems like the special case (and maybe has a better name `GetOrEmpty` or something). Let's punt on that for now.

Reply:
Also maybe we want to run `strings.Lower` on the input here. The user may not know about the case-insensitive nature of 2822

Reply:
We run `strings.ToLower` in slow-path. Hmm, there is missing piece of documentation, that describes using lowercase keys for `Get` as "fast-path". If key is not found then it is lowercased and tried again.
I wrote this by guessing `strings.ToLower(key)` to be much slower operation than `p.values[key]`

Reply:
I don't think we should be worried about speed - I'd just lower everything and do it once to avoid mental overhead when reasoning about the code

Reply:
I don't think we need to worry about speed - it's going to add quite a bit to reason about and mental overhead. Let's lower all keys before looking up, we could leave `Values` exported and tell people to not touch if speed really matters to them. Or perhaps a way to get a copy of the map.